s2spy/traintest.py

```diff
@@ -8,71 +8,173 @@
+        n_intervals = self.traintest_df['i_interval'].max()+1
+        data = self.traintest_df[splits].values[::n_intervals].swapaxes(0,1)
+        self.traintest_xr = xr.DataArray(data=data, 
+                             dims=['split', 'anchor_year'], 
+                             coords=[range(len(splits)), anchor_years])
+        
+    def split_iterate(self, *data_args, y=None):
+        """Train/test iterator.
+
+        Iterator of train/test splits from given data.
+
+        Args:
+            data (xr.Dataset or pd.DataFrame): Dataset with `split` as dimension or
+                DataFrame with `split_no.` as column. E.g. data resampled using
+                `s2spy.time.AdventCalendar`'s `resample` method will have the `key`
+                'anchor_year'.
+        return:
+            Generator contains train data and test data seperately based on splits.
+        """
+        
+        splits = [key for key in self.traintest_df.keys() if 'split' in key]
+        if not splits:
+            raise ValueError("Input data must contain train/test splits."
+                             "Initialize traintest_splits with a splitter and calendar.")        
+        
+        # check the input data type
+        train_Xs = [] ; test_Xs = []
+        for data in data_args:
+            for i, split in enumerate(splits):
+                if isinstance(data, xr.Dataset):
+                    train_X = data.sel(anchor_year = self.traintest_xr.sel(split=i) == "train")
+                    test_X = data.sel(anchor_year = self.traintest_xr.sel(split=i) == "test")
+                elif isinstance(data, pd.DataFrame):
+                    train_X = data.loc[self.traintest_df[f'{split}'] == "train"]
+                    test_X = data.loc[self.traintest_df[f'{split}'] == "test"]
+                else:
+                    raise ValueError("Input data_args should be (list of) type xr.Dataset or pd.DataFrame")
+            train_Xs.append(train_X)
+            test_Xs.append(test_X)
+
+        # if user does not pas list of data_args, it will expect return of input dtype
+        if len(train_Xs) == 1:
+            train_Xs = train_Xs[0]
+            test_Xs = test_Xs[0]
+
+        # check the input y type
+        if y is None:
+            yield train_Xs, test_Xs
+        else:
+            for i, split in enumerate(splits):
+                if isinstance(y, xr.Dataset):
+                    train_y = y.sel(anchor_year = self.traintest_xr.sel(split=i) == "train")
+                    test_y = y.sel(anchor_year = self.traintest_xr.sel(split=i) == "test")
+                elif isinstance(y, pd.DataFrame):
+                    train_y = y.loc[self.traintest_df[f'{split}'] == "train"]
+                    test_y = y.loc[self.traintest_df[f'{split}'] == "test"]
+                else:
+                    raise ValueError("Input y should be of type xr.Dataset or pd.DataFrame")
+
+            yield train_Xs, train_y, test_Xs, test_y
```

Comment:
If I read this code correctly, this will iterate over the input data args, whereas I would expect a loop over the splits.